families/track_and_trade/server/db/records.js

```diff
@@ -0,0 +1,148 @@
+
+const getAttribute = attr => obj => obj(attr)
+const getRecordId = getAttribute('recordId')
+const getName = getAttribute('name')
+const getPublicKey = getAttribute('publicKey')
+const getDataType = getAttribute('dataType')
+const getReporters = getAttribute('reporters')
+const getAuthorization = getAttribute('authorized')
+const getReportedValues = getAttribute('reportedValues')
+
+const getAssociatedAgentId = role => record => record(role).nth(-1)('agentId')
+const getOwnerId = getAssociatedAgentId('owners')
+const getCustodianId = getAssociatedAgentId('custodians')
+
+const hasName = name => obj => r.eq(name, getName(obj))
+const hasRecordId = recordId => obj => r.eq(recordId, getRecordId(obj))
+const hasPublicKey = publicKey => obj => r.eq(publicKey, getPublicKey(obj))
+
+// const hasAttribute = getAttr => attr => obj => r.eq(attr, getAttr(obj))
+// const hasName = hasAttribute(getName)
+// const hasRecordId = hasAttribute(getRecordId)
+
+const hasBlock = block => obj => {
+  return r.and(
+    obj('startBlockNum').le(block),
+    obj('endBlockNum').gt(block)
+  )
+}
+
+const queryWithCurrentBlock = query => {
+  return r.table('blocks')
+    .orderBy(r.desc('blockNum'))
+    .nth(0)('blockNum')
+    .do(query)
+}
+
+const getTable = (tableName, block) =>
+      r.table(tableName).filter(hasBlock(block))
+
+const findRecord = recordId => block => {
+  return getTable('records', block)
+    .filter(hasRecordId(recordId))
+    .nth(0)
+}
+
+const findProperty = recordId => propertyName => block => {
+  return getTable('properties', block)
+    .filter(hasRecordId(recordId))
+    .filter(hasName(propertyName))
+    .nth(0)
+}
+
+const getAuthorizedReporterKeys = property => {
+  return getReporters(property)
+    .filter(getAuthorization)
+    .map(getPublicKey)
+    .coerceTo('array')
+}
+
+const getReporter = publicKey => block => {
+  return getTable('agents', block)
+    .filter(hasPublicKey(publicKey))
+    .pluck('name', 'publicKey')
+    .nth(0)
+}
+
+const findReportedValues =
+      recordId => propertyName => dataType => reporterKeys => block => {
+        return getTable('propertyPages', block)
+          .filter(hasRecordId(recordId))
+          .filter(hasName(propertyName))
+          .concatMap(getReportedValues)
+          .map(getUpdate(dataType)(reporterKeys)(block))
+          .orderBy(r.desc('timestamp'))
+          .coerceTo('array')
+      }
+
+const getValue = dataType => value => {
+  return r.branch(
+    r.eq(dataType, 'INT'), value('intValue'),
+    r.eq(dataType, 'STRING'), value('stringValue'),
+    r.eq(dataType, 'FLOAT'), value('floatValue'),
+    r.eq(dataType, 'BYTES'), value('bytesValue'),
+    r.eq(dataType, 'LOCATION'), value('locationValue'),
+    'intValue'
+  )
+}
+
+const getUpdate = dataType => reporterKeys => block => value => {
+  return r.expr({
+    'value': getValue(dataType)(value),
+    'timestamp': value('timestamp'),
+    'reporter': getReporter(reporterKeys.nth(value('reporterIndex')))(block)
```

Comment:
You don't need quotes around the keys in JS objects unless they include special characters like `-` or ` `. JS will convert them to a string automatically.

Reply:
I come from languages with a clear and sensible distinction between strings and evaluated expressions, so that kind of thing makes me uncomfortable. If it's equivalent, I would prefer to leave the quotation marks.

Reply:
You don't need it, but it's a stylistic difference that fine.